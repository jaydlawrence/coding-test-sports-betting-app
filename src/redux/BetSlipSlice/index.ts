import { createSlice, PayloadAction } from '@reduxjs/toolkit';
import { RootState } from '../../app/store';

export interface BetType {
  eventName: string;
  eventId: string;
  marketName: string;
  marketId: string;
  selectionName: string;
  selectionId: string;
  selectionPrice: number;
}

type BetSlipState = {
  bets: BetType[],
}

const initialState: BetSlipState = {
  bets: [],
};

export const betSlipSlice = createSlice({
  name: 'betSlips',
  initialState,
  reducers: {
    AddRemoveBet: (state, action: PayloadAction<BetType>) => {
      const alreadyExistIndex = state.bets.findIndex( bet => bet.selectionId === action.payload.selectionId);
      
      if (alreadyExistIndex === -1) {
        state.bets.push(action.payload);
      } else {
        state.bets.splice(alreadyExistIndex, 1);
      }
    },
  },
});

export const { AddRemoveBet } = betSlipSlice.actions;

export const selectBetSlipSelectionIds = (state: RootState) => state.betSlips.bets.map( bet => bet.selectionId);

export default betSlipSlice.reducer;
