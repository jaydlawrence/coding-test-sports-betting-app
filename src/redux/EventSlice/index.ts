import { createSlice, PayloadAction } from '@reduxjs/toolkit';
import { AppThunk, RootState } from '../../app/store';
import { ResponseType } from '../../app/ResponseTypes';
import axios from 'axios';

export type EventState = {
  isLoading: boolean,
  response: ResponseType,
  error: string,
}

export const initialState: EventState = {
  isLoading: false,
  response: [],
  error: ''
};

export const eventSlice = createSlice({
  name: 'events',
  initialState,
  reducers: {
    startLoading: state => {
      state.isLoading = true;
      state.error = '';
      state.response = [];
    },
    setError: (state, action: PayloadAction<string> ) => {
      state.isLoading = false;
      state.error = action.payload;
    },
    setEventState: (state, action: PayloadAction<ResponseType>) => {
      state.response = action.payload;
      state.isLoading = false;
      state.error = '';
    },
  },
});

export const { startLoading, setError, setEventState } = eventSlice.actions;

export const getEventsAsync = (): AppThunk => async dispatch => {
  try {
    const response = await axios.get('http://www.mocky.io/v2/59f08692310000b4130e9f71');
    dispatch(setEventState(response.data));
  } catch (e) {
    dispatch(setError('There was an error while fetching event data'));
  }
};

export const selectEventResponse = (state: RootState) => {  
  return state.events.response;
}

export type selectEventLoadingAndErrorsReturnType = {
  isLoading: boolean,
  error: string,
}

export const selectEventLoadingAndErrors = (state: RootState): selectEventLoadingAndErrorsReturnType => {  
  return {
    isLoading: state.events.isLoading,
    error: state.events.error,
  };
}

export default eventSlice.reducer;
