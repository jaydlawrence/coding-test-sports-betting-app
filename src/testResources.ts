import { ResponseType } from "./app/ResponseTypes";
import { BetType } from "./redux/BetSlipSlice";

export const exampleResponse: ResponseType = [
  {
    "id": "EVT_1",
    "name": "Real Madrid vs Barcelona",
    "markets": [
      {
        "id": "MKT_1",
        "name": "Team to Win",
        "selections": [
          {
            "id": "SEL_1",
            "name": "Real Madrid",
            "price": 1.23
          },
          {
            "id": "SEL_2",
            "name": "Barcelona",
            "price": 2.05
          }
        ]
      }
    ]
  },
];

export const exampleBet: BetType = {
  eventName: exampleResponse[0].name,
  eventId: exampleResponse[0].id,
  marketName: exampleResponse[0].markets[0].name,
  marketId: exampleResponse[0].markets[0].id,
  selectionName: exampleResponse[0].markets[0].selections[0].name,
  selectionId: exampleResponse[0].markets[0].selections[0].id,
  selectionPrice: exampleResponse[0].markets[0].selections[0].price,
};
