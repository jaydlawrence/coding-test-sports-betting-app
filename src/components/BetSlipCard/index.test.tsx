import React from 'react';
import configureStore, { MockStore } from 'redux-mock-store';
import renderer from 'react-test-renderer';
import { Provider } from 'react-redux';
import { BetSlipCard } from '.';
import { RemoveBet } from '../../redux/BetSlipSlice';
import { exampleBet } from '../../testResources';
 
const mockStore = configureStore([]);

describe('BetSlipCard', () => {
  let store: MockStore;
  let component: renderer.ReactTestRenderer;

  beforeEach(() => {
    store = mockStore({});
 
    store.dispatch = jest.fn();
 
    component = renderer.create(
      <Provider store={store}>
        <BetSlipCard bet={exampleBet} />
      </Provider>
    );
  });

  it('should display', () => {
    expect(component.toJSON()).toMatchSnapshot();
  });

  it('should fire remove dispatch on remove button click', () => {
    renderer.act(() => {
      component.root.findByType('button').props.onClick();
    });
 
    expect(store.dispatch).toHaveBeenCalledTimes(1);
    expect(store.dispatch).toHaveBeenCalledWith(
      RemoveBet(exampleBet.selectionId)
    );
  });
});