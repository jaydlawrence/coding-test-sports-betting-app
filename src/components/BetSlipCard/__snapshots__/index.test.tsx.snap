// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`BetSlipCard should display 1`] = `
<div
  className="sc-bdfBwQ fosXkd"
>
  <div
    className="MuiPaper-root MuiCard-root sc-gsTCUz fGSzGT MuiPaper-elevation1 MuiPaper-rounded"
  >
    <div
      className="MuiCardHeader-root sc-dlfnbm kqAmLO"
    >
      <div
        className="MuiCardHeader-content"
      >
        <span
          className="MuiTypography-root MuiCardHeader-title MuiTypography-h5 MuiTypography-displayBlock"
        >
          Real Madrid - Team to Win
        </span>
      </div>
    </div>
    <div
      className="MuiCardContent-root sc-hKgILt gzengP"
    >
      <h6
        className="MuiTypography-root sc-eCssSg hqKdWk MuiTypography-h6"
      >
        1.23
      </h6>
      <button
        aria-label="delete"
        className="MuiButtonBase-root MuiIconButton-root"
        disabled={false}
        onBlur={[Function]}
        onClick={[Function]}
        onDragLeave={[Function]}
        onFocus={[Function]}
        onKeyDown={[Function]}
        onKeyUp={[Function]}
        onMouseDown={[Function]}
        onMouseLeave={[Function]}
        onMouseUp={[Function]}
        onTouchEnd={[Function]}
        onTouchMove={[Function]}
        onTouchStart={[Function]}
        tabIndex={0}
        type="button"
      >
        <span
          className="MuiIconButton-label"
        >
          <svg
            aria-hidden={true}
            className="MuiSvgIcon-root MuiSvgIcon-colorSecondary MuiSvgIcon-fontSizeLarge"
            focusable="false"
            viewBox="0 0 24 24"
          >
            <path
              d="M14.59 8L12 10.59 9.41 8 8 9.41 10.59 12 8 14.59 9.41 16 12 13.41 14.59 16 16 14.59 13.41 12 16 9.41 14.59 8zM12 2C6.47 2 2 6.47 2 12s4.47 10 10 10 10-4.47 10-10S17.53 2 12 2zm0 18c-4.41 0-8-3.59-8-8s3.59-8 8-8 8 3.59 8 8-3.59 8-8 8z"
            />
          </svg>
        </span>
      </button>
    </div>
  </div>
</div>
`;
